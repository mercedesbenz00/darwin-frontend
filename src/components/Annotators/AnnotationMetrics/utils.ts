import moment from 'moment'

import { DatasetAnnotationReportPayload, MembershipPayload } from '@/store/types'

import {
  AdaptedDataPoint,
  AggregatedDataPoint,
  AnnotationDataGranularity,
  AnnotationDimensionData,
  AnnotationDataRange
} from './types'

/* builds a moment timestamp formatted a string matching those returned from backend */
export const buildFrom = (range: AnnotationDataRange) =>
  moment
    .utc()
    .subtract(1, range as moment.unitOfTime.DurationConstructor)
    .startOf('hour')
    .format('YYYY-MM-DDTHH:00:00.000000')

/**
 * Takes aggregated source report data and addaptes it for rendering on the line chart.
 *
 * This is done by computing necessary data points.
 */
export const adaptReportData = (r: AggregatedDataPoint): AdaptedDataPoint => ({
  annotationTime: r.annotation_time,
  avgTimePerAnnotation: r.annotations_created > 0
    ? r.annotation_time / r.annotations_created
    : null,
  date: r.date,
  userId: r.user_id,
  imagesAnnotated: r.images_annotated,
  annotationsCreated: r.annotations_created,
  reviewPassRate: r.images_approved > 0
    ? r.images_approved / (r.images_approved + r.images_rejected) * 100
    : null
})

/**
 * Returns a default adapted data point for a team member
 */
export const defaultData = (member: MembershipPayload, date: string): AggregatedDataPoint => ({
  date,
  user_id: member.user_id,
  annotation_time: 0,
  annotations_created: 0,
  annotations_approved: 0,
  images_annotated: 0,
  images_approved: 0,
  images_rejected: 0
})

export const getDimensionsData = (data: AdaptedDataPoint[]): AnnotationDimensionData => {
  const annotationTime = data.reduce((acc, r) => acc + r.annotationTime, 0)
  const annotationsCreated = data.reduce((acc, r) => acc + r.annotationsCreated, 0)
  const avgTimePerAnnotation = annotationsCreated === 0 ? null : annotationTime / annotationsCreated
  const imagesAnnotated = data.reduce((acc, r) => acc + r.imagesAnnotated, 0)

  const totalPassRate = data.reduce((acc, r) => acc + (r.reviewPassRate || 0), 0)
  const periodsWithPassRate = data.filter(r => r.reviewPassRate !== null).length
  const reviewPassRate = totalPassRate / periodsWithPassRate

  return {
    annotationTime,
    annotationsCreated,
    avgTimePerAnnotation,
    imagesAnnotated,
    reviewPassRate
  }
}

/**
 * Expands spars data points so each possible date in the series is covered.
 *
 * This is done by adding a default data point for any missing date
 *
 */
export const expandData = (
  member: MembershipPayload,
  data: AggregatedDataPoint[],
  dates: string[]
): AggregatedDataPoint[] =>
  dates.map(date => data.find(d => d.date === date) || defaultData(member, date))

export const aggregateByDate =
  (date: string, annotationReport: DatasetAnnotationReportPayload[]): AggregatedDataPoint => {
    const initial = {
      annotation_time: 0,
      annotations_created: 0,
      annotations_approved: 0,
      images_annotated: 0,
      images_approved: 0,
      images_rejected: 0
    }

    const reduced = annotationReport.filter(r => r.date === date).reduce((acc, r) => {
      acc.annotation_time += r.annotation_time
      acc.annotations_created += r.annotations_created
      acc.annotations_approved += r.annotations_approved
      acc.images_annotated += r.images_annotated
      acc.images_approved += r.images_approved
      acc.images_rejected += r.images_rejected
      return acc
    }, initial)

    return { ...reduced, date, user_id: null }
  }

export const dateRange =
  (fromDate: string, granularity: AnnotationDataGranularity, format: string) => {
    const from = moment.utc(fromDate)
    const now = moment.utc()
    const dates: string[] = []

    const step = granularity as moment.unitOfTime.DurationConstructor

    const diff = now.diff(from, step)

    for (let i = 0; i <= diff; i++) {
      dates.push(from.format(format))
      from.add(1, step)
    }

    return dates
  }

export const chartFormat = (granularity: AnnotationDataGranularity): string =>
  granularity === 'hour'
    ? 'YYYY-MM-DDTHH:00:00.000000'
    : 'YYYY-MM-DD'

/**
 * selectable data ranges for which we define a "from" parameter when fetching data
 */
const DEFINED_RANGES: AnnotationDataRange[] = ['month', 'week', 'day']

export const granularityFromRange = (range: AnnotationDataRange) => {
  if (range === 'day') { return 'hour' }
  if (range === 'week') { return 'hour' }
  return 'day'
}

/**
 * Computes a "from" param based on the specified range
 *
 * ```
 * 'total' -> null
 * 'month' -> start of hour a month ago
 * 'week' -> start of hour a week ago
 * 'day' -> start of hour a day ago
 * ```
 */
export const fromFromRange = (range: AnnotationDataRange) =>
  DEFINED_RANGES.includes(range) ? buildFrom(range) : null
