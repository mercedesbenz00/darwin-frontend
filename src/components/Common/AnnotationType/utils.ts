import { AnnotationTypeName } from '@/store/types'

/**
 * List of non-meta annotation types. These are basically all types except
 * `measures` and `auto_annotate`.
 *
 * This list is important because these types usually get rendered and are in
 * some way visible to the user, while the remaining two are internal.
 */
export const SORTED_NON_META_TYPES: AnnotationTypeName[] = [
  'attributes',
  'auto_annotate',
  'bounding_box',
  'cuboid',
  'directional_vector',
  'ellipse',
  'graph',
  'inference',
  'instance_id',
  'keypoint',
  'line',
  'link',
  'measures',
  'polygon',
  'polyline',
  'skeleton',
  'string',
  'table',
  'tag',
  'text'
]

export const annotationTypeNameValidator = (value: string): boolean =>
  SORTED_NON_META_TYPES.some(t => t === value)
