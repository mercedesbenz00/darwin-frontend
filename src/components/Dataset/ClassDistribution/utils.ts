import { meanBy } from 'lodash'

import { AnnotationClassPayload } from '@/store/types'
import { DatasetReportClassDistributionPayload } from '@/store/types/DatasetReportPayload'

import { ClassOverallDistribution, DistributionStatus } from './types'

/**
 * Mean should be calculated with distribution of which count is greater than or equal 100
 */
export const getMean = (distribution: DatasetReportClassDistributionPayload[]): number => {
  const above100Distribution = distribution.filter(d => d.count >= 100)
  const mean = meanBy(above100Distribution, 'count')
  return isNaN(mean) ? 0 : mean
}

export const getDistributionStatus = (value: number, mean: number): DistributionStatus => {
  if (value < 30) { return DistributionStatus.VERY_LOW_DATA }
  if (value < 100) { return DistributionStatus.LOW_DATA }
  if (mean > 0 && value > mean * 1.5) { return DistributionStatus.VERY_OVERREPRESENTED }
  if (mean > 0 && value > mean * 1.3) { return DistributionStatus.OVERREPRESENTED }
  if (mean > 0 && value < mean * 0.5) { return DistributionStatus.VERY_UNDERREPRESENTED }
  if (mean > 0 && value < mean * 0.7) { return DistributionStatus.UNDERREPRESENTED }
  if (value < 500) { return DistributionStatus.MEDIUM_LOW_DATA }
  return DistributionStatus.BALANCED
}

export const getAnnotationClassDistribution = (
  annotationClasses: AnnotationClassPayload[],
  distributionByImages: DatasetReportClassDistributionPayload[],
  distributionByInstances: DatasetReportClassDistributionPayload[]
): ClassOverallDistribution[] => {
  const classIds = annotationClasses.map(annotationClass => annotationClass.id)
  const filteredDistributionByInstances = distributionByInstances
    .filter((distribution) => classIds.includes(distribution.id))
  const mean = getMean(filteredDistributionByInstances)

  const classDistributions = annotationClasses.map((annotationClass) => {
    const matchImage = distributionByImages.find((d) => d.id === annotationClass.id)
    const matchInstance = filteredDistributionByInstances.find((d) => d.id === annotationClass.id)
    return {
      annotationClass,
      countByImages: matchImage ? matchImage.count : 0,
      countByInstances: matchInstance ? matchInstance.count : 0,
      status: getDistributionStatus(matchInstance ? matchInstance.count : 0, mean)
    }
  })
  return classDistributions
}
