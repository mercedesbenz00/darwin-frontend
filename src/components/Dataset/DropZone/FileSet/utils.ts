import { FilesByCategory, THUMB_LIMIT } from './types'

/**
 * Returns a list of thumbnails (as data urls), to be rendered for this set.
 *
 * Relies on a watcher to load those thumbnails first.
 */
export const getThumbnailFromFiles = (files: FilesByCategory): (string | ArrayBuffer)[] => {
  const { videos, images, others } = files
  // get 3 frames for each of the first 3 videos
  const thumbsByVideos = videos.slice(0, THUMB_LIMIT).map(video => video.data.thumbs || [])

  // use first frame from each of the first 3 videos
  const thumbs = thumbsByVideos.map(tv => tv[0])

  if (thumbs.length >= THUMB_LIMIT) { return thumbs }

  // less than 3 videos, next, take available frames from images
  images
    .slice(0, THUMB_LIMIT - thumbs.length)
    .forEach(uploadFile => {
      if (uploadFile.data.dataURL) { thumbs.push(uploadFile.data.dataURL) }
    })

  if (thumbs.length >= THUMB_LIMIT) { return thumbs }

  if (videos.length > 0) {
    // limit still not reached, take remaining frames from first video
    const firstVideo = videos[0]
    if (firstVideo.data.thumbs) {
      thumbs.splice(1, 0, ...firstVideo.data.thumbs.slice(1, THUMB_LIMIT - thumbs.length + 1))
    }
  }

  if (thumbs.length >= THUMB_LIMIT) { return thumbs }

  others
    .slice(0, THUMB_LIMIT - thumbs.length)
    .forEach(uploadFile => {
      if (uploadFile.data.dataURL) { thumbs.push(uploadFile.data.dataURL) }
    })

  return thumbs
}
