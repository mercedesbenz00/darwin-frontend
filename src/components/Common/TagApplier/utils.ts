import { AnnotationClassPayload } from '@/store/types'
import { rgbaStringToHSLAString } from '@/utils'

export const tagBrightenedColor = (color: string): string =>
  rgbaStringToHSLAString(color, { l: 70 })

export const matchClassByName = (
  classes: AnnotationClassPayload[],
  text: string
): AnnotationClassPayload | undefined =>
  classes.find(c => c.name === text)
