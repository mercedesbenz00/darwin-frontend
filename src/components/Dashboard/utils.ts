import { MetricPayload } from '@/store/types'
import { RGBA, rgbaString } from '@/utils'

import { GradientLineChartDataset } from './types'

export const metricChartData = (metric: MetricPayload): GradientLineChartDataset[] => {
  const color: RGBA = { r: 0, g: 217, b: 201, a: 1.0 }
  const gradient = { 0: rgbaString(color, 0.3), 0.5: 'rgba(255, 255, 255, 0)' }
  const fillColor = rgbaString(color, 1)

  return [{
    data: metric.data,
    label: metric.name,
    backgroundColor: { gradient: gradient },
    borderColor: fillColor,
    borderWidth: 1,
    pointRadius: 0,
    pointBackgroundColor: fillColor,
    pointBorderWidth: 0
  }]
}
