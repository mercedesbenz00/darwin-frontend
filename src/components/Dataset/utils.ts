import { pluralize } from '@/utils'

export const getBlockedFilesMessage = (blockedFileNames: string[]): string => {
  let strFileNames: string = ''
  const len: number = blockedFileNames.length

  if (len > 5) {
    strFileNames = blockedFileNames.slice(0, 5).join(', ')
    return `Ignoring already uploaded files: ${strFileNames} and ${
      len - 5
    } more ${pluralize(len - 5, 'file', 'files', false)}`
  } else {
    strFileNames = blockedFileNames.join(', ')
    return `Ignoring already uploaded ${pluralize(
      len,
      'file',
      'files',
      false
    )}: ${strFileNames}`
  }
}
