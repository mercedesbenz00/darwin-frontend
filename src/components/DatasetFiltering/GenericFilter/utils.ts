import mime from 'mime'

import { GenericFilterOptionType } from '@/components/DatasetFiltering/GenericFilter'
import {
  DatasetFolderPayload,
  DatasetItemFilenamePayload,
  DatasetItemType,
  DatasetPayload,
  MembershipPayload,
  V2DatasetFolderPayload
} from '@/store/types'
import { getFullName } from '@/utils'

export const membershipToOption = (member: MembershipPayload): GenericFilterOptionType => ({
  id: `annotator-${member.id}`,
  label: getFullName(member),
  type: 'assignees',
  includeHeader: false,
  data: member
})

export const folderToOption = (
  folder: DatasetFolderPayload,
  dataset: DatasetPayload
): GenericFilterOptionType => ({
  id: `folder-${folder.path}`,
  label: folder.path === '/' ? dataset.name : folder.path,
  type: 'paths',
  includeHeader: false,
  data: folder
})

export const folderToOptionV2 = (
  folder: V2DatasetFolderPayload,
  dataset: DatasetPayload
): GenericFilterOptionType => ({
  id: `folder-${folder.path}`,
  label: folder.path === '/' ? dataset.name : folder.path,
  type: 'paths',
  includeHeader: false,
  data: folder
})

export const filenameToOption = (
  item: DatasetItemFilenamePayload
): GenericFilterOptionType => ({
  id: `filename-${item.filename}`,
  label: item.filename,
  type: 'filenames',
  includeHeader: false,
  data: item
})

/**
 * By default, we don't load all the filenames in the initial load.
 * Although if the route query contains the filename,
 * predict the filetype from its filename,
 * so that we can show proper icons in the generic filter.
 */
export const filenameStringToFilenamePayload = (
  filename: string
): DatasetItemFilenamePayload => {
  const data: DatasetItemFilenamePayload = {
    filename,
    type: DatasetItemType.image,
    isDicom: false,
    isPdf: false
  }

  const ext = filename.split('.').pop()
  if (!ext) { return data }

  if (ext.toLowerCase() === 'dcm') {
    data.type = DatasetItemType.playbackVideo
    data.isDicom = true
    return data
  }

  if (ext.toLowerCase() === 'pdf') {
    data.type = DatasetItemType.playbackVideo
    data.isPdf = true
    return data
  }

  const type = mime.getType(ext.trim() || '')
  if (type && type.startsWith('video')) {
    data.type = DatasetItemType.playbackVideo
    return data
  }

  return data
}
