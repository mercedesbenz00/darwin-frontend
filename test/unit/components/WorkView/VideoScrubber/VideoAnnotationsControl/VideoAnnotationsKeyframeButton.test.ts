import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'

import { buildFlaskStageVideoAnnotation } from 'test/unit/components/WorkView/VideoScrubber/utils'
import createTestStore from 'test/unit/createTestStore'
import { stubDirectiveWithAttribute } from 'test/unit/directiveStubs'
import {
  buildAnnotationClassPayload,
  buildDatasetPayload,
  buildLoadedVideo
} from 'test/unit/factories'

import VideoAnnotationsKeyframeButton from '@/components/WorkView/VideoScrubber/VideoAnnotationsControl/VideoAnnotationsKeyframeButton.vue'
import { Editor } from '@/engine/editor'
import { ItemManager } from '@/engine/managers'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.directive('tooltip', stubDirectiveWithAttribute)

let store: ReturnType<typeof createTestStore>
let editor: Editor
let propsData: { editor: Editor }

const sfh = buildDatasetPayload({ id: 1 })
const flask = buildAnnotationClassPayload({
  id: 1,
  name: 'flask',
  annotation_types: ['polygon'],
  datasets: [{ id: sfh.id }]
})

beforeEach(() => {
  store = createTestStore()
  store.commit('aclass/SET_CLASSES', [flask])
  store.commit('workview/SET_DATASET', sfh)
  editor = new Editor(new ItemManager(store), store)
  const annotation = buildFlaskStageVideoAnnotation(editor)
  jest.spyOn(editor.activeView.annotationManager, 'annotations', 'get')
    .mockReturnValue([annotation])
  propsData = { editor }
})

it('matches snapshot when current frame is keyframe', () => {
  editor.activeView.annotations[0].select(false)
  editor.loadedVideo = buildLoadedVideo({ id: 1, currentFrameIndex: 0 })
  const wrapper = shallowMount(VideoAnnotationsKeyframeButton, { localVue, propsData, store })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when current frame is not keyframe', () => {
  editor.activeView.annotations[0].select(false)
  editor.loadedVideo = buildLoadedVideo({ id: 1, currentFrameIndex: 1 })
  const wrapper = shallowMount(VideoAnnotationsKeyframeButton, { localVue, propsData, store })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when no loaded video', () => {
  editor.activeView.annotations[0].select(false)
  const wrapper = shallowMount(VideoAnnotationsKeyframeButton, { localVue, propsData, store })
  expect(wrapper).toMatchSnapshot()
})

it('removes current selected keyframe on click if current frame is keyframe', async () => {
  editor.activeView.annotations[0].select(false)
  editor.loadedVideo = buildLoadedVideo({ id: 1, currentFrameIndex: 0 })
  jest.spyOn(editor.activeView, 'deleteKeyFrame').mockReturnValue(undefined)
  const wrapper = shallowMount(VideoAnnotationsKeyframeButton, { localVue, propsData, store })
  wrapper.find('button').trigger('click')
  await wrapper.vm.$nextTick()
  expect(editor.activeView.deleteKeyFrame).toBeCalledWith(editor.activeView.annotations[0], 0)
})

it('creates a new keyframe on click if current frame is not keyframe', async () => {
  editor.activeView.annotations[0].select(false)
  editor.loadedVideo = buildLoadedVideo({ id: 1, currentFrameIndex: 1 })
  jest.spyOn(editor.activeView, 'createKeyFrame').mockReturnValue(undefined)
  const wrapper = shallowMount(VideoAnnotationsKeyframeButton, { localVue, propsData, store })
  wrapper.find('button').trigger('click')
  await wrapper.vm.$nextTick()
  expect(editor.activeView.createKeyFrame).toBeCalledWith(editor.activeView.annotations[0], 1)
})
