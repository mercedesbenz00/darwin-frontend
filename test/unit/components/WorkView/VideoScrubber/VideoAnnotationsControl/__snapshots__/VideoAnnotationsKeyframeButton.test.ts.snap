// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot when current frame is keyframe 1`] = `
<button class="keyframe-button keyframe-button--selected" tooltip="{'content':'Remove current keyframe <span class=\\'tooltip__hotkey\\'>I</span>','placement':'top','delay':1000}">
  <keyframe-icon-stub></keyframe-icon-stub>
</button>
`;

exports[`matches snapshot when current frame is not keyframe 1`] = `
<button class="keyframe-button" tooltip="{'content':'Create a new keyframe <span class=\\'tooltip__hotkey\\'>I</span>','placement':'top','delay':1000}">
  <keyframe-icon-stub></keyframe-icon-stub>
</button>
`;

exports[`matches snapshot when no loaded video 1`] = ``;
